clang/lib/CIR/Dialect/Transforms/LoweringPrepareAArch64CXXABI.cpp

```diff
@@ -0,0 +1,258 @@
+  // registers of the appropriate kind.
+  auto regSizeValue = builder.create<mlir::cir::ConstantOp>(
+      loc, regOffs.getType(),
+      mlir::cir::IntAttr::get(regOffs.getType(), regSize));
+  auto newOffset = builder.create<mlir::cir::BinOp>(
+      loc, regOffs.getType(), mlir::cir::BinOpKind::Add, regOffs, regSizeValue);
+  builder.createStore(loc, newOffset, regOffsP);
+  // Now we're in a position to decide whether this argument really was in
+  // registers or not.
+  auto inRegs = builder.create<mlir::cir::CmpOp>(
+      loc, boolTy, mlir::cir::CmpOpKind::le, newOffset, zeroValue);
+  builder.create<mlir::cir::BrCondOp>(loc, inRegs, inRegBlock, onStackBlock);
+
+  //=======================================
+  // Argument was in registers
+  //=======================================
+  // Now we emit the code for if the argument was originally passed in
+  // registers. First start the appropriate block:
+  builder.setInsertionPointToEnd(inRegBlock);
+  auto regTopP = builder.createGetMemberOp(
+      loc, valist, IsFPR ? "vr_top" : "gr_top", regTopIndex);
+  auto regTop = builder.create<mlir::cir::LoadOp>(loc, regTopP);
+  auto i8Ty = mlir::IntegerType::get(builder.getContext(), 8);
+  auto i8PtrTy = mlir::cir::PointerType::get(builder.getContext(), i8Ty);
+  auto castRegTop = builder.createBitcast(regTop, i8PtrTy);
+  // On big-endian platforms, the value will be right-aligned in its stack slot.
+  // and we also need to think about other ABI lowering concerns listed below.
+  assert(!cir::MissingFeatures::handleBigEndian());
+  assert(!cir::MissingFeatures::handleAArch64Indirect());
+  assert(!cir::MissingFeatures::supportisHomogeneousAggregateQueryForAArch64());
+  assert(!cir::MissingFeatures::supportTySizeQueryForAArch64());
+  assert(!cir::MissingFeatures::supportTyAlignQueryForAArch64());
+
+  auto resAsInt8P = builder.create<mlir::cir::PtrStrideOp>(
+      loc, castRegTop.getType(), castRegTop, regOffs);
+  auto resAsVoidP = builder.createBitcast(resAsInt8P, regTop.getType());
+  builder.create<mlir::cir::BrOp>(loc, mlir::ValueRange{resAsVoidP}, contBlock);
+
+  //=======================================
+  // Argument was on the stack
+  //=======================================
+  builder.setInsertionPointToEnd(onStackBlock);
+  auto stackP = builder.createGetMemberOp(loc, valist, "stack", 0);
+
+  auto onStackPtr = builder.create<mlir::cir::LoadOp>(loc, stackP);
+  auto ptrDiffTy =
+      mlir::cir::IntType::get(builder.getContext(), 64, /*signed=*/false);
+
+  // On big-endian platforms, the value will be right-aligned in its stack slot
+  // Also, the consideration involves type size and alignment, arg indirectness
+  // which are all ABI defined thus need ABI lowering query system.
+  // The implementation we have now supports most common cases which assumes
+  // no indirectness, no alignment greater than 8, and little endian.
+  assert(!cir::MissingFeatures::handleBigEndian());
```

Comment:
I rather see the big endian check in the place where it happens in the original code (this should be applied elsewhere in this function where big endian is queried), skeleton is important! Also, `CIRDataLayout` does support big endian queriers, so you can query it and assert on the leg that's not implemented.

Reply:
lib/CIR/CodeGen/CIRDataLayout.h:22:class CIRDataLayout
The problem (again) is CIRDataLayout is declared and defined in CodeGen dir. 
So I just added another missing feature.

Reply:
Oh right, as part of LoweringPrepare we gonna have to keep adding these bits as we go. Sorry for the detour, but please create another PR that moves `CIRDataLayout` to `include/clang/CIR/Dialect/IR/`, should be straight fwd. We land that and you rebase this one on top so you can use it here.